{
 "cells": [
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "First we made database from 16s sequences downloaded from Variovorax, Polaromonas complete genomes and Herbidospora (incomplete assemblies):\n",
    "\n",
    "makeblastdb -in combined.fasta -out Bacteria_database -dbtype nucl -title \"database\" -parse_seqids\n",
    "\n",
    "Then we ran the blastn command with e_value 1e-18:\n",
    "blastn -db Bacteria_database -query sequences.fna -evalue 1e-18 -out alignments.outfmt6 -outfmt 6\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#reading alignment table\n",
    "data = pd.read_csv(\"/home/is6/ASVs_comparison/Blast_results/alignments.outfmt6\", sep=\"\\t\", header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# necessary functions\n",
    "# explode a dataframe column\n",
    "def explode(df, lst_cols, fill_value='', preserve_index=False):\n",
    "    # make sure `lst_cols` is list-alike\n",
    "    if (lst_cols is not None\n",
    "        and len(lst_cols) > 0\n",
    "        and not isinstance(lst_cols, (list, tuple, np.ndarray, pd.Series))):\n",
    "        lst_cols = [lst_cols]\n",
    "    # all columns except `lst_cols`\n",
    "    idx_cols = df.columns.difference(lst_cols)\n",
    "    # calculate lengths of lists\n",
    "    lens = df[lst_cols[0]].str.len()\n",
    "    # preserve original index values    \n",
    "    idx = np.repeat(df.index.values, lens)\n",
    "    # create \"exploded\" DF\n",
    "    res = (pd.DataFrame({\n",
    "                col:np.repeat(df[col].values, lens)\n",
    "                for col in idx_cols},\n",
    "                index=idx)\n",
    "             .assign(**{col:np.concatenate(df.loc[lens>0, col].values)\n",
    "                            for col in lst_cols}))\n",
    "    # append those rows that have empty lists\n",
    "    if (lens == 0).any():\n",
    "        # at least one list in cells is empty\n",
    "        res = (res.append(df.loc[lens==0, idx_cols], sort=False)\n",
    "                  .fillna(fill_value))\n",
    "    # revert the original index order\n",
    "    res = res.sort_index()\n",
    "    # reset index if requested\n",
    "    if not preserve_index:        \n",
    "        res = res.reset_index(drop=True)\n",
    "    return res\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Set column name to the table\n",
    "data.columns = ['query_id', 'subject_id', 'query_identity', 'alignment_length', 'mismatch', 'gap_open', 'qstart', 'qend', 'sstart','send', 'e_value', 'bitscore']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query_id</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>query_identity</th>\n",
       "      <th>alignment_length</th>\n",
       "      <th>mismatch</th>\n",
       "      <th>gap_open</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>sstart</th>\n",
       "      <th>send</th>\n",
       "      <th>e_value</th>\n",
       "      <th>bitscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>seq3</td>\n",
       "      <td>GCA.28_010093065.1</td>\n",
       "      <td>94.862</td>\n",
       "      <td>253</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>253</td>\n",
       "      <td>527</td>\n",
       "      <td>779</td>\n",
       "      <td>1.610000e-112</td>\n",
       "      <td>396</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>seq3</td>\n",
       "      <td>GCA.27_010093065.1</td>\n",
       "      <td>94.862</td>\n",
       "      <td>253</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>253</td>\n",
       "      <td>527</td>\n",
       "      <td>779</td>\n",
       "      <td>1.610000e-112</td>\n",
       "      <td>396</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>seq3</td>\n",
       "      <td>GCA.26_010093065.1</td>\n",
       "      <td>94.862</td>\n",
       "      <td>253</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>253</td>\n",
       "      <td>527</td>\n",
       "      <td>779</td>\n",
       "      <td>1.610000e-112</td>\n",
       "      <td>396</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>seq3</td>\n",
       "      <td>GCA.25_010093065.1</td>\n",
       "      <td>94.862</td>\n",
       "      <td>253</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>253</td>\n",
       "      <td>527</td>\n",
       "      <td>779</td>\n",
       "      <td>1.610000e-112</td>\n",
       "      <td>396</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>seq3</td>\n",
       "      <td>GCA.24_010093065.1</td>\n",
       "      <td>94.862</td>\n",
       "      <td>253</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>253</td>\n",
       "      <td>527</td>\n",
       "      <td>779</td>\n",
       "      <td>1.610000e-112</td>\n",
       "      <td>396</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  query_id          subject_id  query_identity  alignment_length  mismatch  \\\n",
       "0     seq3  GCA.28_010093065.1          94.862               253        13   \n",
       "1     seq3  GCA.27_010093065.1          94.862               253        13   \n",
       "2     seq3  GCA.26_010093065.1          94.862               253        13   \n",
       "3     seq3  GCA.25_010093065.1          94.862               253        13   \n",
       "4     seq3  GCA.24_010093065.1          94.862               253        13   \n",
       "\n",
       "   gap_open  qstart  qend  sstart  send        e_value  bitscore  \n",
       "0         0       1   253     527   779  1.610000e-112       396  \n",
       "1         0       1   253     527   779  1.610000e-112       396  \n",
       "2         0       1   253     527   779  1.610000e-112       396  \n",
       "3         0       1   253     527   779  1.610000e-112       396  \n",
       "4         0       1   253     527   779  1.610000e-112       396  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# reading taxa files as dataframes\n",
    "sxt_df = pd.read_csv(\"/home/is6/16s_preprocessing/ksenia/taxonomy.tsv\", sep='\\t', header=None)\n",
    "\n",
    "sxt_df.columns = ['query_id', 'root', 'kingdom', 'phylum', 'order', 'class', 'family', 'genus', 'species']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query_id</th>\n",
       "      <th>root</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>order</th>\n",
       "      <th>class</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>seq1</td>\n",
       "      <td>Root</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Bacillales</td>\n",
       "      <td>Bacillaceae</td>\n",
       "      <td>Bacillus</td>\n",
       "      <td>unclassified_Bacillus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>seq2</td>\n",
       "      <td>Root</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Bacillales</td>\n",
       "      <td>Bacillaceae</td>\n",
       "      <td>Bacillus</td>\n",
       "      <td>unclassified_Bacillus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>seq3</td>\n",
       "      <td>Root</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteroidota</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>Chitinophagales</td>\n",
       "      <td>Chitinophagaceae</td>\n",
       "      <td>unclassified_Chitinophagaceae</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>seq4</td>\n",
       "      <td>Root</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteroidota</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>Cytophagales</td>\n",
       "      <td>Microscillaceae</td>\n",
       "      <td>unclassified_Microscillaceae</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>seq5</td>\n",
       "      <td>Root</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Azospirillales</td>\n",
       "      <td>Inquilinaceae</td>\n",
       "      <td>Inquilinus</td>\n",
       "      <td>unclassified_Inquilinus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  query_id  root   kingdom          phylum                order  \\\n",
       "0     seq1  Root  Bacteria      Firmicutes              Bacilli   \n",
       "1     seq2  Root  Bacteria      Firmicutes              Bacilli   \n",
       "2     seq3  Root  Bacteria    Bacteroidota          Bacteroidia   \n",
       "3     seq4  Root  Bacteria    Bacteroidota          Bacteroidia   \n",
       "4     seq5  Root  Bacteria  Proteobacteria  Alphaproteobacteria   \n",
       "\n",
       "             class            family                          genus  \\\n",
       "0       Bacillales       Bacillaceae                       Bacillus   \n",
       "1       Bacillales       Bacillaceae                       Bacillus   \n",
       "2  Chitinophagales  Chitinophagaceae  unclassified_Chitinophagaceae   \n",
       "3     Cytophagales   Microscillaceae   unclassified_Microscillaceae   \n",
       "4   Azospirillales     Inquilinaceae                     Inquilinus   \n",
       "\n",
       "                   species  \n",
       "0    unclassified_Bacillus  \n",
       "1    unclassified_Bacillus  \n",
       "2                      NaN  \n",
       "3                      NaN  \n",
       "4  unclassified_Inquilinus  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sxt_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query_id</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>query_identity</th>\n",
       "      <th>alignment_length</th>\n",
       "      <th>mismatch</th>\n",
       "      <th>gap_open</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>sstart</th>\n",
       "      <th>send</th>\n",
       "      <th>e_value</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>root</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>order</th>\n",
       "      <th>class</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>seq3</td>\n",
       "      <td>GCA.28_010093065.1</td>\n",
       "      <td>94.862</td>\n",
       "      <td>253</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>253</td>\n",
       "      <td>527</td>\n",
       "      <td>779</td>\n",
       "      <td>1.610000e-112</td>\n",
       "      <td>396</td>\n",
       "      <td>Root</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteroidota</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>Chitinophagales</td>\n",
       "      <td>Chitinophagaceae</td>\n",
       "      <td>unclassified_Chitinophagaceae</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>seq3</td>\n",
       "      <td>GCA.27_010093065.1</td>\n",
       "      <td>94.862</td>\n",
       "      <td>253</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>253</td>\n",
       "      <td>527</td>\n",
       "      <td>779</td>\n",
       "      <td>1.610000e-112</td>\n",
       "      <td>396</td>\n",
       "      <td>Root</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteroidota</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>Chitinophagales</td>\n",
       "      <td>Chitinophagaceae</td>\n",
       "      <td>unclassified_Chitinophagaceae</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>seq3</td>\n",
       "      <td>GCA.26_010093065.1</td>\n",
       "      <td>94.862</td>\n",
       "      <td>253</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>253</td>\n",
       "      <td>527</td>\n",
       "      <td>779</td>\n",
       "      <td>1.610000e-112</td>\n",
       "      <td>396</td>\n",
       "      <td>Root</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteroidota</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>Chitinophagales</td>\n",
       "      <td>Chitinophagaceae</td>\n",
       "      <td>unclassified_Chitinophagaceae</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>seq3</td>\n",
       "      <td>GCA.25_010093065.1</td>\n",
       "      <td>94.862</td>\n",
       "      <td>253</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>253</td>\n",
       "      <td>527</td>\n",
       "      <td>779</td>\n",
       "      <td>1.610000e-112</td>\n",
       "      <td>396</td>\n",
       "      <td>Root</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteroidota</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>Chitinophagales</td>\n",
       "      <td>Chitinophagaceae</td>\n",
       "      <td>unclassified_Chitinophagaceae</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>seq3</td>\n",
       "      <td>GCA.24_010093065.1</td>\n",
       "      <td>94.862</td>\n",
       "      <td>253</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>253</td>\n",
       "      <td>527</td>\n",
       "      <td>779</td>\n",
       "      <td>1.610000e-112</td>\n",
       "      <td>396</td>\n",
       "      <td>Root</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteroidota</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>Chitinophagales</td>\n",
       "      <td>Chitinophagaceae</td>\n",
       "      <td>unclassified_Chitinophagaceae</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  query_id          subject_id  query_identity  alignment_length  mismatch  \\\n",
       "0     seq3  GCA.28_010093065.1          94.862               253        13   \n",
       "1     seq3  GCA.27_010093065.1          94.862               253        13   \n",
       "2     seq3  GCA.26_010093065.1          94.862               253        13   \n",
       "3     seq3  GCA.25_010093065.1          94.862               253        13   \n",
       "4     seq3  GCA.24_010093065.1          94.862               253        13   \n",
       "\n",
       "   gap_open  qstart  qend  sstart  send        e_value  bitscore  root  \\\n",
       "0         0       1   253     527   779  1.610000e-112       396  Root   \n",
       "1         0       1   253     527   779  1.610000e-112       396  Root   \n",
       "2         0       1   253     527   779  1.610000e-112       396  Root   \n",
       "3         0       1   253     527   779  1.610000e-112       396  Root   \n",
       "4         0       1   253     527   779  1.610000e-112       396  Root   \n",
       "\n",
       "    kingdom        phylum        order            class            family  \\\n",
       "0  Bacteria  Bacteroidota  Bacteroidia  Chitinophagales  Chitinophagaceae   \n",
       "1  Bacteria  Bacteroidota  Bacteroidia  Chitinophagales  Chitinophagaceae   \n",
       "2  Bacteria  Bacteroidota  Bacteroidia  Chitinophagales  Chitinophagaceae   \n",
       "3  Bacteria  Bacteroidota  Bacteroidia  Chitinophagales  Chitinophagaceae   \n",
       "4  Bacteria  Bacteroidota  Bacteroidia  Chitinophagales  Chitinophagaceae   \n",
       "\n",
       "                           genus species  \n",
       "0  unclassified_Chitinophagaceae     NaN  \n",
       "1  unclassified_Chitinophagaceae     NaN  \n",
       "2  unclassified_Chitinophagaceae     NaN  \n",
       "3  unclassified_Chitinophagaceae     NaN  \n",
       "4  unclassified_Chitinophagaceae     NaN  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# merge 16S dataframe on 'seq' to blast results dataframe to get annotations\n",
    "j_data = pd.merge(data, sxt_df, on=\"query_id\")\n",
    "j_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#filtering cover 100% of ASV sequences and 100 identity to reference sequences\n",
    "N_data = j_data[(j_data[['query_identity']] >= 100).all(axis=1)]\n",
    "N_data = N_data[(N_data[['mismatch']] == 0).all(axis=1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "28"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(N_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query_id</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>query_identity</th>\n",
       "      <th>alignment_length</th>\n",
       "      <th>mismatch</th>\n",
       "      <th>gap_open</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>sstart</th>\n",
       "      <th>send</th>\n",
       "      <th>...</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>root</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>order</th>\n",
       "      <th>class</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>nsseqid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>203</th>\n",
       "      <td>seq15</td>\n",
       "      <td>GCA.21_000024005.1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>253</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>253</td>\n",
       "      <td>520</td>\n",
       "      <td>772</td>\n",
       "      <td>...</td>\n",
       "      <td>468</td>\n",
       "      <td>Root</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteroidota</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>Chitinophagales</td>\n",
       "      <td>Chitinophagaceae</td>\n",
       "      <td>Chitinophaga</td>\n",
       "      <td>unclassified_Chitinophaga</td>\n",
       "      <td>GCA_000024005.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204</th>\n",
       "      <td>seq15</td>\n",
       "      <td>GCA.20_000024005.1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>253</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>253</td>\n",
       "      <td>520</td>\n",
       "      <td>772</td>\n",
       "      <td>...</td>\n",
       "      <td>468</td>\n",
       "      <td>Root</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteroidota</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>Chitinophagales</td>\n",
       "      <td>Chitinophagaceae</td>\n",
       "      <td>Chitinophaga</td>\n",
       "      <td>unclassified_Chitinophaga</td>\n",
       "      <td>GCA_000024005.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205</th>\n",
       "      <td>seq15</td>\n",
       "      <td>GCA.19_000024005.1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>253</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>253</td>\n",
       "      <td>520</td>\n",
       "      <td>772</td>\n",
       "      <td>...</td>\n",
       "      <td>468</td>\n",
       "      <td>Root</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteroidota</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>Chitinophagales</td>\n",
       "      <td>Chitinophagaceae</td>\n",
       "      <td>Chitinophaga</td>\n",
       "      <td>unclassified_Chitinophaga</td>\n",
       "      <td>GCA_000024005.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>206</th>\n",
       "      <td>seq15</td>\n",
       "      <td>GCA.18_000024005.1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>253</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>253</td>\n",
       "      <td>520</td>\n",
       "      <td>772</td>\n",
       "      <td>...</td>\n",
       "      <td>468</td>\n",
       "      <td>Root</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteroidota</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>Chitinophagales</td>\n",
       "      <td>Chitinophagaceae</td>\n",
       "      <td>Chitinophaga</td>\n",
       "      <td>unclassified_Chitinophaga</td>\n",
       "      <td>GCA_000024005.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207</th>\n",
       "      <td>seq15</td>\n",
       "      <td>GCA.17_000024005.1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>253</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>253</td>\n",
       "      <td>520</td>\n",
       "      <td>772</td>\n",
       "      <td>...</td>\n",
       "      <td>468</td>\n",
       "      <td>Root</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteroidota</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>Chitinophagales</td>\n",
       "      <td>Chitinophagaceae</td>\n",
       "      <td>Chitinophaga</td>\n",
       "      <td>unclassified_Chitinophaga</td>\n",
       "      <td>GCA_000024005.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    query_id          subject_id  query_identity  alignment_length  mismatch  \\\n",
       "203    seq15  GCA.21_000024005.1           100.0               253         0   \n",
       "204    seq15  GCA.20_000024005.1           100.0               253         0   \n",
       "205    seq15  GCA.19_000024005.1           100.0               253         0   \n",
       "206    seq15  GCA.18_000024005.1           100.0               253         0   \n",
       "207    seq15  GCA.17_000024005.1           100.0               253         0   \n",
       "\n",
       "     gap_open  qstart  qend  sstart  send  ...  bitscore  root   kingdom  \\\n",
       "203         0       1   253     520   772  ...       468  Root  Bacteria   \n",
       "204         0       1   253     520   772  ...       468  Root  Bacteria   \n",
       "205         0       1   253     520   772  ...       468  Root  Bacteria   \n",
       "206         0       1   253     520   772  ...       468  Root  Bacteria   \n",
       "207         0       1   253     520   772  ...       468  Root  Bacteria   \n",
       "\n",
       "           phylum        order            class            family  \\\n",
       "203  Bacteroidota  Bacteroidia  Chitinophagales  Chitinophagaceae   \n",
       "204  Bacteroidota  Bacteroidia  Chitinophagales  Chitinophagaceae   \n",
       "205  Bacteroidota  Bacteroidia  Chitinophagales  Chitinophagaceae   \n",
       "206  Bacteroidota  Bacteroidia  Chitinophagales  Chitinophagaceae   \n",
       "207  Bacteroidota  Bacteroidia  Chitinophagales  Chitinophagaceae   \n",
       "\n",
       "            genus                    species          nsseqid  \n",
       "203  Chitinophaga  unclassified_Chitinophaga  GCA_000024005.1  \n",
       "204  Chitinophaga  unclassified_Chitinophaga  GCA_000024005.1  \n",
       "205  Chitinophaga  unclassified_Chitinophaga  GCA_000024005.1  \n",
       "206  Chitinophaga  unclassified_Chitinophaga  GCA_000024005.1  \n",
       "207  Chitinophaga  unclassified_Chitinophaga  GCA_000024005.1  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# seperating sseqid column into gca and s-id, then renaming GCAxx-\n",
    "# format into just GCA and combining the columns back together\n",
    "# to get the original id as nsseqid column\n",
    "\n",
    "N_data[['gca', 's_id']] = N_data[\"subject_id\"].str.split(\"_\", n = 1, expand = True) \n",
    "N_data['gca'] = 'GCA'\n",
    "N_data['nsseqid'] = N_data[['gca', 's_id']].agg('_'.join, axis=1)\n",
    "\n",
    "N_data = N_data.drop('gca', 1)\n",
    "N_data = N_data.drop('s_id', 1)\n",
    "N_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject_taxonomy</th>\n",
       "      <th>nsseqid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Herbidospora</td>\n",
       "      <td>GCA_005233835.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Herbidospora</td>\n",
       "      <td>GCA_009901565.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Herbidospora</td>\n",
       "      <td>GCA_001570525.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Herbidospora</td>\n",
       "      <td>GCA_001570605.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Herbidospora</td>\n",
       "      <td>GCA_001570585.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  subject_taxonomy          nsseqid\n",
       "0     Herbidospora  GCA_005233835.1\n",
       "1     Herbidospora  GCA_009901565.1\n",
       "2     Herbidospora  GCA_001570525.1\n",
       "3     Herbidospora  GCA_001570605.1\n",
       "4     Herbidospora  GCA_001570585.1"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# getting all sample annotations as a dataframe\n",
    "dir_path = \"/home/is6/ASVs_comparison/samples\"\n",
    "dir_name = []\n",
    "file_name = []\n",
    "for path, dirs,files in os.walk(dir_path):\n",
    "    dir_name.append(dirs)\n",
    "    file_name.append(files)\n",
    "    \n",
    "new_file_list = list(filter(None,file_name))\n",
    "new_dir_list = list(dir_name[0])\n",
    "all_samples = list(zip(new_dir_list,new_file_list))\n",
    "\n",
    "sample_df = pd.DataFrame(all_samples)\n",
    "sample_df.columns = [\"genera\", \"samples\"]\n",
    "\n",
    "sample_df['samples'] = sample_df['samples'].astype(str)\n",
    "new_df = pd.DataFrame(explode(sample_df.assign(samples=sample_df.samples.str.split(',')), 'samples'))\n",
    "\n",
    "new_df.samples = new_df.samples.str.replace(\"'\", '')\n",
    "new_df.samples = new_df.samples.str.replace(\"[\", '')\n",
    "new_df.samples = new_df.samples.str.replace(\"]\", '')\n",
    "new_df.samples = new_df.samples.str.replace(\" \", '')\n",
    "\n",
    "new_df.samples = new_df[\"samples\"].str.rsplit(\"_\", n = 2, expand = True)\n",
    "new_df.columns = ['subject_taxonomy', 'nsseqid']\n",
    "\n",
    "new_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject_taxonomy</th>\n",
       "      <th>nsseqid</th>\n",
       "      <th>query_id</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>query_identity</th>\n",
       "      <th>alignment_length</th>\n",
       "      <th>mismatch</th>\n",
       "      <th>gap_open</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>...</th>\n",
       "      <th>e_value</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>root</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>order</th>\n",
       "      <th>class</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Herbidospora</td>\n",
       "      <td>GCA_005233835.1</td>\n",
       "      <td>seq62</td>\n",
       "      <td>GCA.32_005233835.1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>253</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>253</td>\n",
       "      <td>...</td>\n",
       "      <td>3.360000e-134</td>\n",
       "      <td>468</td>\n",
       "      <td>Root</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteriota</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Streptosporangiales</td>\n",
       "      <td>Streptosporangiaceae</td>\n",
       "      <td>unclassified_Streptosporangiaceae</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Sorangium</td>\n",
       "      <td>GCA_004135755.1</td>\n",
       "      <td>seq32</td>\n",
       "      <td>GCA.52_004135755.1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>253</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>253</td>\n",
       "      <td>...</td>\n",
       "      <td>3.360000e-134</td>\n",
       "      <td>468</td>\n",
       "      <td>Root</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Myxococcota</td>\n",
       "      <td>Polyangia</td>\n",
       "      <td>Polyangiales</td>\n",
       "      <td>Polyangiaceae</td>\n",
       "      <td>Sorangium</td>\n",
       "      <td>unclassified_Sorangium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Sorangium</td>\n",
       "      <td>GCA_004135755.1</td>\n",
       "      <td>seq32</td>\n",
       "      <td>GCA.51_004135755.1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>253</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>253</td>\n",
       "      <td>...</td>\n",
       "      <td>3.360000e-134</td>\n",
       "      <td>468</td>\n",
       "      <td>Root</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Myxococcota</td>\n",
       "      <td>Polyangia</td>\n",
       "      <td>Polyangiales</td>\n",
       "      <td>Polyangiaceae</td>\n",
       "      <td>Sorangium</td>\n",
       "      <td>unclassified_Sorangium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Sorangium</td>\n",
       "      <td>GCA_004135755.1</td>\n",
       "      <td>seq32</td>\n",
       "      <td>GCA.50_004135755.1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>253</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>253</td>\n",
       "      <td>...</td>\n",
       "      <td>3.360000e-134</td>\n",
       "      <td>468</td>\n",
       "      <td>Root</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Myxococcota</td>\n",
       "      <td>Polyangia</td>\n",
       "      <td>Polyangiales</td>\n",
       "      <td>Polyangiaceae</td>\n",
       "      <td>Sorangium</td>\n",
       "      <td>unclassified_Sorangium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Sorangium</td>\n",
       "      <td>GCA_004135755.1</td>\n",
       "      <td>seq32</td>\n",
       "      <td>GCA.49_004135755.1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>253</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>253</td>\n",
       "      <td>...</td>\n",
       "      <td>3.360000e-134</td>\n",
       "      <td>468</td>\n",
       "      <td>Root</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Myxococcota</td>\n",
       "      <td>Polyangia</td>\n",
       "      <td>Polyangiales</td>\n",
       "      <td>Polyangiaceae</td>\n",
       "      <td>Sorangium</td>\n",
       "      <td>unclassified_Sorangium</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  subject_taxonomy          nsseqid query_id          subject_id  \\\n",
       "0     Herbidospora  GCA_005233835.1    seq62  GCA.32_005233835.1   \n",
       "1        Sorangium  GCA_004135755.1    seq32  GCA.52_004135755.1   \n",
       "2        Sorangium  GCA_004135755.1    seq32  GCA.51_004135755.1   \n",
       "3        Sorangium  GCA_004135755.1    seq32  GCA.50_004135755.1   \n",
       "4        Sorangium  GCA_004135755.1    seq32  GCA.49_004135755.1   \n",
       "\n",
       "   query_identity  alignment_length  mismatch  gap_open  qstart  qend  ...  \\\n",
       "0           100.0               253         0         0       1   253  ...   \n",
       "1           100.0               253         0         0       1   253  ...   \n",
       "2           100.0               253         0         0       1   253  ...   \n",
       "3           100.0               253         0         0       1   253  ...   \n",
       "4           100.0               253         0         0       1   253  ...   \n",
       "\n",
       "         e_value  bitscore  root   kingdom            phylum           order  \\\n",
       "0  3.360000e-134       468  Root  Bacteria  Actinobacteriota  Actinobacteria   \n",
       "1  3.360000e-134       468  Root  Bacteria       Myxococcota       Polyangia   \n",
       "2  3.360000e-134       468  Root  Bacteria       Myxococcota       Polyangia   \n",
       "3  3.360000e-134       468  Root  Bacteria       Myxococcota       Polyangia   \n",
       "4  3.360000e-134       468  Root  Bacteria       Myxococcota       Polyangia   \n",
       "\n",
       "                 class                family  \\\n",
       "0  Streptosporangiales  Streptosporangiaceae   \n",
       "1         Polyangiales         Polyangiaceae   \n",
       "2         Polyangiales         Polyangiaceae   \n",
       "3         Polyangiales         Polyangiaceae   \n",
       "4         Polyangiales         Polyangiaceae   \n",
       "\n",
       "                               genus                 species  \n",
       "0  unclassified_Streptosporangiaceae                     NaN  \n",
       "1                          Sorangium  unclassified_Sorangium  \n",
       "2                          Sorangium  unclassified_Sorangium  \n",
       "3                          Sorangium  unclassified_Sorangium  \n",
       "4                          Sorangium  unclassified_Sorangium  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# merge final dataframe on 'sseqid' to 16s_extracted dataframe to get species name\n",
    "final_data = pd.merge(new_df,N_data, on=\"nsseqid\")\n",
    "final_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "28"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(final_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query_id</th>\n",
       "      <th>nsseqid</th>\n",
       "      <th>mismatch</th>\n",
       "      <th>query_identity</th>\n",
       "      <th>subject_taxonomy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>seq62</td>\n",
       "      <td>GCA_005233835.1</td>\n",
       "      <td>0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>Herbidospora</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>seq32</td>\n",
       "      <td>GCA_004135755.1</td>\n",
       "      <td>0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>Sorangium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>seq32</td>\n",
       "      <td>GCA_004135755.1</td>\n",
       "      <td>0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>Sorangium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>seq32</td>\n",
       "      <td>GCA_004135755.1</td>\n",
       "      <td>0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>Sorangium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>seq32</td>\n",
       "      <td>GCA_004135755.1</td>\n",
       "      <td>0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>Sorangium</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  query_id          nsseqid  mismatch  query_identity subject_taxonomy\n",
       "0    seq62  GCA_005233835.1         0           100.0     Herbidospora\n",
       "1    seq32  GCA_004135755.1         0           100.0        Sorangium\n",
       "2    seq32  GCA_004135755.1         0           100.0        Sorangium\n",
       "3    seq32  GCA_004135755.1         0           100.0        Sorangium\n",
       "4    seq32  GCA_004135755.1         0           100.0        Sorangium"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#query_id subject_id query_coverage query_identity subject_taxonomy\n",
    "\n",
    "final_data = final_data[['query_id','nsseqid', 'mismatch', 'query_identity','subject_taxonomy']]\n",
    "final_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query_id</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>mismatch/query_coverage</th>\n",
       "      <th>query_identity</th>\n",
       "      <th>subject_taxonomy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>seq62</td>\n",
       "      <td>GCA_005233835.1</td>\n",
       "      <td>0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>Herbidospora</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>seq32</td>\n",
       "      <td>GCA_004135755.1</td>\n",
       "      <td>0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>Sorangium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>seq32</td>\n",
       "      <td>GCA_004135755.1</td>\n",
       "      <td>0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>Sorangium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>seq32</td>\n",
       "      <td>GCA_004135755.1</td>\n",
       "      <td>0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>Sorangium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>seq32</td>\n",
       "      <td>GCA_004135755.1</td>\n",
       "      <td>0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>Sorangium</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  query_id       subject_id  mismatch/query_coverage  query_identity  \\\n",
       "0    seq62  GCA_005233835.1                        0           100.0   \n",
       "1    seq32  GCA_004135755.1                        0           100.0   \n",
       "2    seq32  GCA_004135755.1                        0           100.0   \n",
       "3    seq32  GCA_004135755.1                        0           100.0   \n",
       "4    seq32  GCA_004135755.1                        0           100.0   \n",
       "\n",
       "  subject_taxonomy  \n",
       "0     Herbidospora  \n",
       "1        Sorangium  \n",
       "2        Sorangium  \n",
       "3        Sorangium  \n",
       "4        Sorangium  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_data = final_data.rename(columns={'mismatch':'mismatch/query_coverage','nsseqid':'subject_id'})\n",
    "final_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "final_data.to_csv (\"/home/is6/ASVs_comparison/16s_database/BLAST_results.csv\", index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
