---
title: "Denoising of 16s rRNA data provided for cellulase project"
output: pdf_document
---

The file contains preprocessing steps for 16s rRNA amplicon data. It includes DADA2 pipeline and taxonomy annotation. Primer excision was done by cutadapt before this steps. DADA2 performed with pseudo-pooling. After inferring, merging (2695 sequences) and chimera removing we got 2174 sequences. Taxonomy prediction was done by Idtaxa with 80% confidence and gave us 322 sequences in the end of preprocessing. We left taxa identified at least at genus level.


```{r message=FALSE, warning=FALSE}
library(dada2)
library(phyloseq)
library(DECIPHER)
library(dplyr)
library(ggplot2)
library(reshape2)
library(svglite)
library(tidyr)


set.seed(130)

setwd("//home/is7/cellulase/")

n.threads = 30

```

Extract sample names

Filter and trim reads

```{r}
root.raw <- './noprimer'
forward.raw <- sort(list.files(root.raw, pattern='R1_001.trimmed.fastq.gz', full.names = TRUE))
reverse.raw <- sort(list.files(root.raw, pattern='R2_001.trimmed.fastq.gz', full.names = TRUE))
sample.names <- sapply(strsplit(basename(forward.raw), "_"), `[`, 1)
```

Inspect quality

```{r}
# Forward reads
plotQualityProfile(forward.raw, aggregate=TRUE, n=100000)

# Reverse reads
plotQualityProfile(reverse.raw, aggregate=TRUE, n=100000)
```

Filter and trim reads

```{r}
root.qc = './qc'
forward.qc <- file.path(root.qc, paste0(sample.names, "R1_001.fastq.gz"))
reverse.qc <- file.path(root.qc, paste0(sample.names, "R2_001.fastq.gz"))

qc.out <- filterAndTrim(forward.raw, forward.qc, reverse.raw, reverse.qc, 
                        truncLen=c(250, 190), maxEE=c(3,3), compress=TRUE, multithread=n.threads)

head(qc.out)
```

```{r}
# Forward reads
plotQualityProfile(forward.qc, aggregate=TRUE, n=100000)

# Reverse reads
plotQualityProfile(reverse.qc, aggregate=TRUE, n=100000)
```

Learn the error rates

```{r}
# Forward reads
forward.err <- learnErrors(forward.qc, multithread=n.threads)

# Reverse reads
reverse.err <- learnErrors(reverse.qc, multithread=n.threads)
```

Dereplicate reads

```{r}
forward.derep <- derepFastq(forward.qc, n=1e7, verbose=TRUE)
reverse.derep <- derepFastq(reverse.qc, n=1e7, verbose=TRUE)
# Name the derep-class objects by the sample names
names(forward.derep) <- sample.names
names(reverse.derep) <- sample.names
```

Sample Inference

```{r}
forward.dada <- dada(forward.derep, err=forward.err, pool='pseudo', multithread=n.threads)
reverse.dada <- dada(reverse.derep, err=reverse.err, pool='pseudo', multithread=n.threads)
```

Merge paired reads

```{r}
merged <- mergePairs(forward.dada, forward.derep, reverse.dada, reverse.derep)
```

Build a sequence table and remove chimera
```{r}
sequence.table <- makeSequenceTable(merged)
sequence.table.nochim <- removeBimeraDenovo(sequence.table, method="consensus", multithread=n.threads, verbose=TRUE)
```

Summarise preprocessing statistics
```{r}
get.n <- function(x) sum(getUniques(x))
statistics <- cbind(qc.out, 
               sapply(forward.dada, get.n), 
               sapply(reverse.dada, get.n), 
               sapply(merged, get.n), 
               rowSums(sequence.table.nochim))
colnames(statistics) <- c('noprimer_pairs', 'qc_pairs', 'forward_denoised', 'reverse_denoised', 'merged', 'non-chimeric')
rownames(statistics) <- sample.names

write.table(statistics, file = "statistics.tsv", row.names=FALSE, sep="\t")
```


```{r}
# extract results
seqs <- getSequences(sequence.table.nochim)
seq.names <- paste0("seq", seq(length(seqs)))
names(seqs) <- seq.names

count.table <- otu_table(sequence.table.nochim, taxa_are_rows=FALSE) %>% t
rownames(count.table) <- seq.names

# subset results
freq.mask <- (count.table >= 3) %>% 
    rowSums %>% 
    `/`(ncol(count.table)) %>%
    `>=`(0.25)

seqs.subset <- DNAStringSet(seqs[freq.mask], use.names=TRUE)
count.table.subset <- count.table[freq.mask]

cat('Spared', length(seqs.subset), 'out of', length(seqs), 'sequences\n')
cat('Spared', sum(count.table.subset), 'out of', sum(count.table), 'observations\n')
```

Predict taxonomy
```{r}
classifier <- readRDS('taxonomy/classifier_v4.RData')
predicted.taxonomy <- IdTaxa(seqs.subset, classifier, threshold=80, processors=n.threads)
format.taxonomy <- function(id) paste(id$taxon, sep='', collapse=';')
tax.ids <- sapply(predicted.taxonomy, format.taxonomy)

#write.table(as.data.frame(tax.ids, row.names=names(tax.ids)), 'taxonomy.tsv', #sep='\t', quote=FALSE, col.names=FALSE)

```
Cleaning taxonomy table

Leaving sequences with identified ranks at genus level
```{r}
# separate table into column with corresponded ranks
taxonomy <- as.data.frame(tax.ids, row.names=names(tax.ids)) %>% 
  separate(., col = tax.ids, into = c('root', 'domain', 'phylum', 'class', 'order', 'family', 'genus', 'species'), sep = ';')

# leave only that rows which have information up to genus level
taxonomy <- taxonomy[complete.cases(taxonomy$genus), ]

write.table(taxonomy, 'taxonomy.tsv', sep='\t', quote=FALSE, col.names=FALSE)
```

```{r}

colnames(statistics) <- c('Primer excisition', 'Quality control', 'Denoising (R1)', 'Denoising (R2)', 'Pair merging', 'Chimera elimination')
statistics.long <- melt(statistics)

plot.stats <- ggplot(statistics.long, aes(x=Var2, y=value / 10^3)) + 
    geom_boxplot() +
    scale_x_discrete(limits = rev(levels(statistics.long$Var2))) +
    ylab(bquote('Sequence count, '*10^3*'')) +
    coord_flip() +
    theme_bw() +
    theme(text=element_text(size=11,  family='Helvetica'),
          axis.title.y=element_blank())


ggsave(file='plots/stats.png', plot=plot.stats, width=6, height=2)
```
